Smt/Tactic/WHNFSmt.lean

```diff
@@ -0,0 +1,129 @@
+import Smt.Tactic.WHNFConfigurable
+import Lean.Elab.Term
+import Lean.Elab.Binders
+
+namespace Smt
+
+open Lean
+
+/-- Constants which SMT knows about and we thus don't want to unfold. -/
+def smtConsts : Std.HashSet Name :=
+  List.foldr (fun c s => s.insert c) Std.HashSet.empty
+  [
+    ``Eq,
+    ``BEq.beq,
+    ``bne,
+    ``true,
+    ``false,
+    ``And,
+    ``Or,
+    ``Not,
+    ``ite,
+    ``dite,
+    ``Exists,
+    ``HAdd.hAdd,
+    ``HSub.hSub,
+    ``HMul.hMul,
+    ``HDiv.hDiv,
+    ``HMod.hMod,
+    ``HPow.hPow,
+    ``HAppend.hAppend,
+    ``HAnd.hAnd,
+    ``HXor.hXor,
+    ``HOr.hOr,
+    ``HShiftLeft.hShiftLeft,
+    ``HShiftRight.hShiftRight,
+    ``LE.le,
+    ``LT.lt,
+    ``GE.ge,
+    ``GT.gt,
+    ``and,
+    ``or,
+    ``not,
+    `BitVec,
+    `BitVec.zero,
+    `BitVec.ofNat,
+    `BitVec.extract,
+    /- Eager WHNF unfolds implicit arguments in a way that interacts badly with the projection
+    unfolding in `smt`. For example, `@Xor.xor (BitVec n) (instXorBitVec n)` goes to
+    `@Xor.xor (BitVec n) { val := ... : Fin (2^n) }` and then we get `{ .. : Fin (2^n) }`
+    as the instance. To make projection unfolding work, we prefer `@Xor.xor (BitVec n) BitVec.xor`,
+    so the extra symbols are blocked here. This is not great, so maybe:
+    - investigate using `explicitOnly` in `whnfCore`?
+    - stop unfolding projections in `smt`?
```

Comment:
I am still in favor of unfolding projections, but I don't think unfolding them during translation is a good idea. I was thinking of introducing a tactic that only unfolds projections. Would that help? Another option is to unfold them with your partial evaluator help.

Reply:
The choice of when and what to unfold in general is really tricky. The only hard rule is that we never want to unfold SMT builtins, so from this perspective having a long list here makes some sense. A better solution might be to allow `smtTranslator`s to register which builtins they translate, and query that global `HashSet` before invoking the evaluator, as well as to replace the `knownConsts` list.

As for splitting projection unfolding from the translators, that might indeed make proof reconstruction a bit easier. But since the convention in general is to keep things like arithmetic operators folded (e.g. `add_comm (a b : MyFoo) : a + b = b + a` is a better theorem than `add_comm (a b) : MyFoo.add a b = MyFoo.add b a`), could you remind me why we want to unfold projections?

Reply:
The original reason for unfolding projections is to simplify the translation. For translation, we simply cannot keep all projections folded. Actually, the only ones we can keep folded are the ones that we (the developers) know correspond to SMT builtin symbols. So, at some point, we will have to unfold the ones we don't know about (e.g., `a + b` to `MyFoo.add a b`) since we do not know which SMT symbol they correspond to. In that case, wouldn't it be simpler to unfold all projections? I do not see what unfolding some projections but not the others will save us...

BTW, I am not saying that the `smt` tactic should reject `add_comm (a b : MyFoo) : a + b = b + a` (it should accept and prove both). I am only suggesting that the core tactic (translation + proof reconstruction) should only deal with `add_comm (a b) : MyFoo.add a b = MyFoo.add b a`. It's then up to the plugins to lift the proof to `add_comm (a b : MyFoo) : a + b = b + a`.

Reply:
Okay, I can see this. In particular if a user defines `MyFoo := Nat` with `instance : Add MyFoo where add a b := (a : Nat) + (b : Nat)`, we need to unfold it to know how to translate it. Indeed doing this in a preprocessing tactic rather than during translation might be better. It is actually a special case of specialization.